### ISSUE TITLE
Inconsistency in array description

### ISSUE THREAD

**Issue by goganchic** [apiaryio/drafter#557]
I've been looking through array description section in MSON spec and I've found [such section](https://github.com/apiaryio/mson/blob/master/MSON%20Specification.md#351-type-specification). From this section I've found that if I define such array: `array[number, string]` it means that this structure can contain numbers and strings. So I think that such array is valid `['a', 1, 'b', 'c']` and such [1, 2, 3] and such `['a', 'b', 'c']`. Array can contain any number of elements but all of them should be either a string or a number.

It if run Drafter on such document:

```
# My API

# GET /foo

+ Response 200 (application/json)
    + Attributes (array[number, string], fixed-type)
```

I can see that generated json schema looks so:

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "array",
  "items": [
    {
      "type": "number"
    },
    {
      "type": "string"
    }
  ]
}
```

It means that array should contain two or more elements. First element should be a number and second should be a string. It's completely different from MSON spec. Is it a bug or a feature?

I expect to have such json schema:

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "array",
  "items": {
    "anyOf": [
      {"type": "number"},
      {"type": "string"}
    ]
  }
}
```

Bigger question is describing two types of arrays:

* array containing fixed number of elements of fixed types
* array containing arbitrary number of elements each belongs to some set of types

I think it's better to have separate type for describing array containing fixed number of elements of fixed types. In several programming languages (like python or erlang) special type `tuple` exists for such task. If we have separate type in MSON for such task it will be much easier to distinguish these two cases. E.g. `array[number, string]` means array which contains any number of elements each should be a number or a string and `tuple[number, string]` means array which contains two elements first is number and second is a string.

The same idea for in-place elements definition:

```
+ foo (array, fixed-type)
    + Properties
        + a
        + b
    + Properties
        + c
        + d
```

means that array can contain any number of objects of first or second schema and

```
+ foo (tuple)
    + Properties
        + a
        + b
    + Properties
        + c
        + d
```

means that array should contain only two elements first - of first schema and second of second schema.

The case which we have discussed with @kylef [in slack](https://apiblueprint.slack.com/archives/C0E59EQ0P/p1529069446000135) became easier and more consistent if I want to describe array contains any number of strings I write `array[string]` and if I want to describe array contains only one string I write `tuple[string]`.

What do you think about such idea?

**Comment by pksunkara** [comment]
@tjanc You might be interested in this.

**Comment by tjanc** [comment]
First, thanks for the nitpicking. We need far more of that.

> From this section I've found that if I define such array: `array[number, string]` it means that this structure can contain numbers and strings. So I think that such array is valid `['a', 1, 'b', 'c']` and such [1, 2, 3] and such `['a', 'b', 'c']`. Array can contain any number of elements but all of them should be either a string or a number.

You are right. [This section](https://github.com/apiaryio/mson/blob/master/MSON%20Specification.md#43-nested-member-types) on the other hand implies an `array` of a union of types (`number` and `string` in your example) will be denoted by a `fixed-type` attribute. This definition is what I'm working with.

> I can see that generated json schema looks so: [...] Is it a bug or a feature?
It's a bug, and it will be fixed with #552 in drafter.

> I expect to have such json schema: [...]
Apart from drafter rendering JSON Schema draft-4, this is exactly what #552 renders.

> I think it's better to have separate type for describing array containing fixed number of elements of fixed types.
Agreed. It's funny we both came up with the same type "tuple" independently. That adds some validation. Fact is that MSON has a type system that is not very algebraic. For that reason I'm currently doing some research into type systems ([Intuitionistic type theory](https://en.wikipedia.org/wiki/Intuitionistic_type_theory) and [Hindley-Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system)) that will result in a whitepaper for future versions of both MSON and API Elements. *Introducing a `tuple` type however seems straight-forward enough to be implemented before such restructuring will take place.*

If you want to push MSON `tuple` forward, it would be best to open a `t4: feature` issue on the [MSON Spec](https://github.com/apiaryio/mson/issues) project. I would consider the [first definition you referenced](https://github.com/apiaryio/mson/blob/master/MSON%20Specification.md#351-type-specification) as a bug in the MSON Spec.

**Comment by kylef** [comment]
As of current master (6e1a3e31) the example document provided will now produce the following schema:

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "array",
  "items": {
    "anyOf": [
      {
        "type": "number"
      },
      {
        "type": "string"
      }
    ]
  }
}
```

As for the second example:

```javascript
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "foo": {
      "type": "array",
      "items": {
        "anyOf": [
          {
            "type": "object",
            "properties": {
              "a": {
                "type": "string"
              },
              "b": {
                "type": "string"
              }
            }
          },
          {
            "type": "object",
            "properties": {
              "c": {
                "type": "string"
              },
              "d": {
                "type": "string"
              }
            }
          }
        ]
      }
    }
  }
}
```

With minor alterations to the MSON syntax to make it valid:

```mson
# My API

# GET /foo

+ Response 200 (application/json)
    + Attributes
        + foo (array, fixed-type)
            + (object)
                + Properties
                    + a
                    + b
            + (object)
                + Properties
                    + c
                    + d
```
